pinot-controller/src/test/java/org/apache/pinot/controller/helix/core/realtime/segment/SegmentFlushThresholdComputerTest.java

```diff
@@ -0,0 +1,314 @@
+        streamConfig, committingSegmentDescriptor, committingSegmentZKMetadata,
+        partitionGroupMetadataList, "events3__0__0__20211222T1646Z");
+
+    // (totalDocs / segmentSize) * flushThresholdSegmentSize
+    // (30000 / 250000) * 300000
+    assertEquals(threshold, 36_000);
+  }
+
+  @Test
+  public void testNoRows() {
+    SegmentFlushThresholdComputer computer = new SegmentFlushThresholdComputer();
+
+    PartitionLevelStreamConfig streamConfig = mock(PartitionLevelStreamConfig.class);
+    when(streamConfig.getFlushThresholdSegmentSizeBytes()).thenReturn(300_0000L);
+
+    CommittingSegmentDescriptor committingSegmentDescriptor = mock(CommittingSegmentDescriptor.class);
+    when(committingSegmentDescriptor.getSegmentSizeBytes()).thenReturn(250_0000L);
+
+    SegmentZKMetadata committingSegmentZKMetadata = mock(SegmentZKMetadata.class);
+    when(committingSegmentZKMetadata.getTotalDocs()).thenReturn(0L);
+    when(committingSegmentZKMetadata.getSizeThresholdToFlushSegment()).thenReturn(0);
+
+    List<PartitionGroupMetadata> partitionGroupMetadataList = new ArrayList<>();
+    int threshold = computer.computeThreshold(
+        streamConfig, committingSegmentDescriptor, committingSegmentZKMetadata,
+        partitionGroupMetadataList, "events3__0__0__20211222T1646Z");
+
+    // max((totalDocs / segmentSize) * flushThresholdSegmentSize, 10000)
+    // max(0, 10000)
+    assertEquals(threshold, 10_000);
+  }
+
+  @Test
+  public void testAdjustRowsToSizeRatio() {
+    SegmentFlushThresholdComputer computer = new SegmentFlushThresholdComputer();
+
+    PartitionLevelStreamConfig streamConfig = mock(PartitionLevelStreamConfig.class);
+    when(streamConfig.getFlushThresholdSegmentSizeBytes()).thenReturn(300_0000L);
+
+    CommittingSegmentDescriptor committingSegmentDescriptor = mock(CommittingSegmentDescriptor.class);
+    when(committingSegmentDescriptor.getSegmentSizeBytes()).thenReturn(200_000L);
+
+    SegmentZKMetadata committingSegmentZKMetadata = mock(SegmentZKMetadata.class);
+    when(committingSegmentZKMetadata.getTotalDocs()).thenReturn(30_000L, 50_000L);
+    when(committingSegmentZKMetadata.getSizeThresholdToFlushSegment()).thenReturn(60_000);
+
+    List<PartitionGroupMetadata> partitionGroupMetadataList = new ArrayList<>();
+
+    computer.computeThreshold(streamConfig, committingSegmentDescriptor, committingSegmentZKMetadata,
+        partitionGroupMetadataList, "events3__0__0__20211222T1646Z");
+
+    // (totalDocs / segmentSize)
+    // (30000 / 200000)
+    assertEquals(computer.getLatestSegmentRowsToSizeRatio(), 0.15);
+
+    computer.computeThreshold(streamConfig, committingSegmentDescriptor, committingSegmentZKMetadata,
+        partitionGroupMetadataList, "events3__0__0__20211222T1646Z");
+
+    // (0.1 * (totalDocs / segmentSize)) + (0.9 * lastRatio)
+    // (0.1 * (50000 / 200000)) + (0.9 * 0.15)
+    // (0.1 * 0.25) + (0.9 * 0.15)
+    assertEquals(computer.getLatestSegmentRowsToSizeRatio(), 0.16);
+  }
+}
```

Comment:
nit: newline at end of file (there's an intellij setting which adds this automatically)